src/lib/transaction_logic/test/zkapp_logic.ml

```diff
@@ -819,5 +955,200 @@ let%test_module "Test transaction logic." =
                       ] )
                  && Predicates.verify_balances_unchanged ~ledger ~txn accounts )
             )
-            (run_zkapp_cmd ~fee_payer ~fee ~accounts txns) )
+            (run_zkapp_cmd ~global_slot ~fee_payer ~fee ~accounts txns) )
+
+    let%test_unit "Update preconditions can block update application." =
+      Quickcheck.test ~trials
+        (let open Quickcheck.Generator.Let_syntax in
+        let%bind sender =
+          Test_account.gen_constrained_balance
+            ~min:Balance.(of_mina_int_exn 5)
+            ()
+        in
+        let%bind receiver = Test_account.gen_empty in
+        let max_fee =
+          let open Fee in
+          balance_to_fee sender.balance - one |> Option.value ~default:one
+        in
+        let%bind fee = Fee.(gen_incl one max_fee) in
+        let balance_after_fee =
+          let open Balance in
+          sender.balance - Amount.of_fee fee
+          |> Option.value_map ~default:Amount.zero ~f:to_amount
+        in
+        let%bind amount = Amount.(gen_incl one balance_after_fee) in
+        let%bind valid_while_intv = gen_valid_while in
+        let%map global_slot =
+          global_slot_gen_excl valid_while_intv.lower valid_while_intv.upper
+        in
+        let preconditions =
+          let open Account_update in
+          let open Preconditions in
+          { network = Zkapp_precondition.Protocol_state.accept
+          ; account = Account_precondition.Accept
+          ; valid_while = Check valid_while_intv
+          }
+        in
+        let txn = Simple_txn.make ~preconditions ~sender ~receiver amount in
+        ( global_slot
+        , sender.pk
+        , fee
+        , [ sender; receiver ]
+        , [ (txn :> transaction) ] ))
+        ~f:(fun (global_slot, fee_payer, fee, accounts, txns) ->
+          [%test_pred: Zk_cmd_result.t Or_error.t]
+            (Predicates.pure ~f:(fun (txn, ledger) ->
+                 Transaction_status.equal txn.command.status
+                   (Failed
+                      [ []
+                      ; [ Valid_while_precondition_unsatisfied ]
+                      ; [ Valid_while_precondition_unsatisfied ]
+                      ] )
+                 && Predicates.verify_balances_unchanged ~ledger ~txn accounts )
+            )
+            (run_zkapp_cmd ~global_slot ~fee_payer ~fee ~accounts txns) )
+
+    let%test_unit "Account preconditions are checked against the local state." =
+      Quickcheck.test ~trials
+        (let open Quickcheck.Generator.Let_syntax in
+        let%bind sender =
```

Comment:
I cannot see a duplication big enough to justify a helper function. I mean 1 or even 2 consecutive duplicated lines, which are specific to these two tests only, don't justify the extraction, because it's difficult to descriptively name a function which generates a very specific setup. It'll be easier for the reader to comprehend when they read the code itself rather than some name, which won't be descriptive enough or exceedingly long. It would make sense, though, if there were a couple of lines doing a job which can be named concisely or if there was a biogger chunk of duplicated code, but as I said, I cannot see one. If you can point me to what you mean exactly, then I'm happy to extract.

Reply:
I would extract calculation of a tuple of `sender, receiver, amount, fee` and have a function like `gen_test_payment_data`. I'll leave this decision to you though.